osu.Game/Database/ArchiveModelManager.cs

```diff
@@ -130,56 +132,50 @@ protected ArchiveModelManager(Storage storage, IDatabaseContextFactory contextFa
         /// This will post notifications tracking progress.
         /// </summary>
         /// <param name="paths">One or more archive locations on disk.</param>
-        public void Import(params string[] paths)
+        public async Task Import(params string[] paths)
         {
             var notification = new ProgressNotification { State = ProgressNotificationState.Active };
 
             PostNotification?.Invoke(notification);
-            Import(notification, paths);
+
+            await Import(notification, paths);
         }
 
-        protected void Import(ProgressNotification notification, params string[] paths)
+        protected async Task Import(ProgressNotification notification, params string[] paths)
         {
             notification.Progress = 0;
             notification.Text = "Import is initialising...";
 
             var term = $"{typeof(TModel).Name.Replace("Info", "").ToLower()}";
 
-            List<TModel> imported = new List<TModel>();
-
             int current = 0;
 
-            foreach (string path in paths)
+            var imported = new List<TModel>();
+
+            await Task.WhenAll(paths.Select(path => Import(path, notification.CancellationToken).ContinueWith(t =>
             {
-                if (notification.State == ProgressNotificationState.Cancelled)
-                    // user requested abort
+                if (notification.CancellationToken.IsCancellationRequested)
                     return;
 
-                try
+                lock (notification)
```

Comment:
`notification` is an external resource from this class' point of view. And since "one should only put a lock on a resource only she knows about", I recommend changing the lock here to another object. I think it should be a simple `private readonly object notificationLock = new object();` among fields.

I think this is important.

Reply:
This is temporary code and not intended to be in the final PR.